File: googletest/test/gtest_rgt_output_test_.cc
// Copyright (C) 2024 Sony Interactive Entertainment Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Sony Interactive Entertainment Inc. nor the
// names of its contributors may be used to endorse or promote products
// derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

// The purpose of this file is to generate Google Test output for various
// "rotten green test" conditions. In a "rotten" test, the test passes, but
// at least one test assertion was not executed. The output will then be
// verified by gtest_rgt_output_test.py to ensure that Google Test generates
// the desired messages. Therefore, many tests in this file are MEANT TO BE
// ROTTEN.
//
// However, NONE ARE MEANT TO FAIL. If the test has an overall failing result,
// the helper methods won't have their rotten assertions reported, and that is
// something we do want to verify. Instead, googletest-ouput-test shows that
// (a) an overall Fail result means that rotten helper methods aren't reported;
// (b) failing assertions are not reported as rotten.
//
// This test shows (c) rotten assertions (and rotten helpers) in a passing test
// are reported properly.
//
// Note: "assertions" here can mean either EXPECT_* or ASSERT_*.
//
// Modeled on googletest-output-test.

#include "gtest/gtest.h"
#include "src/gtest-internal-inl.h"

#if GTEST_HAS_RGT

// Tests that will not be reported as rotten for various reasons.

TEST(NotRotten, EmptyIsntRotten) { }

TEST(NotRotten, DISABLED_DisabledIsntRotten) {
  EXPECT_EQ(0, 1);
}

TEST(NotRotten, SkippedIsntRotten) {
  if (testing::internal::AlwaysTrue()) {
    GTEST_SKIP();
  }
  EXPECT_EQ(1, 2);
}

TEST(NotRotten, AllPass) {
  EXPECT_EQ(0, 0);
  EXPECT_EQ(1, 1);
}

// EXPECT_[NON]FATAL_FAILURE executes an assertion that is supposed to fail.
// The bookkeeping for assertions will show the failing one as not executed,
// but because it's containined in a Test that passes, we shouldn't report it.

TEST(NotRotten, ExpectNonfatalFailureIsntRotten) {
  EXPECT_NONFATAL_FAILURE({ EXPECT_EQ(1, 0) << "Non-fatal"; }, "Non-fatal");
}

TEST(NotRotten, ExpectFatalFailureIsntRotten) {
  EXPECT_FATAL_FAILURE( { ASSERT_EQ(1, 0) << "Fatal"; }, "Fatal");
}

// ASSERT/EXPECT_NO_FATAL_FAILURE that is executed (pass or fail) isn't rotten.

void MustPass() { }

TEST(NotRotten, ExpectNoFatalFailurePasses) {
  EXPECT_NO_FATAL_FAILURE(MustPass());
  ASSERT_NO_FATAL_FAILURE(MustPass());
}

// As an unfortunate consequence of how EXPECT_[NON]FATAL_FAILURE is handled,
// if one is executed, it disables rotten detection entirely within the
// containing Test. It would be nice to fix that someday.
TEST(NotRotten, ExpectNonfatalFailureSadlyDisablesRotten) {
  EXPECT_NONFATAL_FAILURE( { EXPECT_EQ(1, 0) << "Non-fatal"; }, "Non-fatal");
  if (testing::internal::AlwaysFalse()) {
    EXPECT_EQ(1, 0);
  }
}

TEST(NotRotten, ExpectFatalFailureSadlyDisablesRotten) {
  EXPECT_FATAL_FAILURE( { ASSERT_EQ(1, 0) << "Fatal"; }, "Fatal");
  if (testing::internal::AlwaysFalse()) {
    EXPECT_EQ(1, 0);
  }
}

// Tests that will be reported as rotten.

TEST(IsRotten, SimpleExpect) {
  if (testing::internal::AlwaysFalse()) {
    EXPECT_EQ(0, 0);
  }
}

TEST(IsRotten, SimpleAssert) {
  if (testing::internal::AlwaysFalse()) {
    ASSERT_EQ(0, 0);
  }
}

TEST(IsRotten, OneIsEnough) {
  EXPECT_EQ(0, 0);
  if (testing::internal::AlwaysFalse()) {
    ASSERT_EQ(1, 1);
  }
}

// If EXPECT_[NON]FATAL_FAILURE exists, but isn't executed, that is detected.
TEST(IsRotten, MissedExpectFailure) {
  if (testing::internal::AlwaysFalse()) {
    EXPECT_NONFATAL_FAILURE( { EXPECT_EQ(1, 0) << "Non-fatal"; }, "Non-fatal");
    EXPECT_FATAL_FAILURE( { ASSERT_EQ(1, 0) << "Fatal"; }, "Fatal");
  }
}

TEST(IsRotten, ExpectNoFatalFailure) {
  if (testing::internal::AlwaysFalse()) {
    EXPECT_NO_FATAL_FAILURE(MustPass());
    ASSERT_NO_FATAL_FAILURE(MustPass());
  }
}

void RottenHelperNeverCalled() {
  EXPECT_EQ(0, 0);
}

// Test that RGT detection works correctly with fixtures.
// Just a few sample repeats, not everything.

class NotRottenFixture : public testing::Test {};
class IsRottenFixture : public testing::Test {};

TEST_F(NotRottenFixture, AllPass) {
  EXPECT_EQ(0, 0);
  ASSERT_EQ(1, 1);
}

TEST_F(NotRottenFixture, SkippedIsntRotten) {
  if (testing::internal::AlwaysTrue()) {
    GTEST_SKIP();
  }
  ASSERT_EQ(0, 1);
}

TEST_F(IsRottenFixture, SingleExpect) {
  if (testing::internal::AlwaysFalse()) {
    EXPECT_EQ(0, 0);
  }
}

// RGT and Parameterized Tests.

// TEST_P has one broad issue with respect to RGT.
//
// Rotten test detection relies on having compile-time access to the TestInfo*
// for each test. TEST_P doesn't have that. As a consequence, in RGT these
// tests behave like helper methods. Lacking any better information, RGT
// reporting assumes that all helper assertions should be executed.
// Unfortunately this means the usual things that disable rotten detection
// (disabled, skipped, EXPECT_[NON]FATAL_FAILURE) are ineffective.
// These tests will therefore be prone to false positives.
//
// However, that behavior is consistent, so we include it in the output test.

class NotRottenParamTest : public testing::TestWithParam<int> {};

TEST_P(NotRottenParamTest, Passes) { EXPECT_GE(2, GetParam()); }

// If the assertions are conditional on the parameter, they aren't all executed
// on each iteration. Because we treat these like helpers, we report them at
// the end of the test, and if every path was eventually taken, we're fine.
// If we support parameterized tests properly in the future, we should continue
// not reporting until all iterations have executed, to avoid false positives.

TEST_P(NotRottenParamTest, ConditionalOnParam) {
  if (GetParam() == 1) {
    EXPECT_EQ(1, 1);
  } else {
    EXPECT_EQ(2, 2);
  }
}

INSTANTIATE_TEST_SUITE_P(, NotRottenParamTest, testing::Values(1, 2));

class IsRottenParamTest : public testing::TestWithParam<int> {};

TEST_P(IsRottenParamTest, SkippedIsSadlyRotten) {
  if (testing::internal::AlwaysTrue()) {
    GTEST_SKIP();
  }
  EXPECT_EQ(0, GetParam());
}

TEST_P(IsRottenParamTest, DISABLED_DisabledIsSadlyRotten) {
  EXPECT_EQ(0, GetParam());
}

TEST_P(IsRottenParamTest, ActuallyRotten) {
  switch (GetParam()) {
  case 0:
    EXPECT_EQ(0, 0);
    break;
  case 1:
    EXPECT_EQ(1, 1);
    break;
  default:
    EXPECT_EQ(2, 2);
    break;
  }
}

INSTANTIATE_TEST_SUITE_P(, IsRottenParamTest, testing::Values(1, 2));

// RGT and Typed Tests.

// TYPED_TEST and TYPED_TEST_P have two broad issues with respect to RGT.
//
// 1) Rotten test detection relies on having compile-time access to the
//    TestInfo* for each test. TYPED_TEST[_P] don't have that. As a
//    consequence, in RGT they behave like helper methods. Lacking any
//    better information, RGT reporting assumes that all helper assertions
//    should be executed. Unfortunately this means the usual things that
//    disable rotten detection (disabled, skipped, EXPECT_[NON]FATAL_FAILURE)
//    are ineffective.
//    These tests will therefore be prone to false positives.
//
// 2) Assertion bookkeeping relies on certain static data being allocated to
//    particular object-file sections. GCC prior to GCC 14 (and sometimes,
//    even in GCC 14) does not obey section attributes for static data in a
//    template (or 'inline') function.
//    See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94342.
//    TYPED_TEST[_P] are implemented as templates, so the assertions info is
//    effectively lost with older GCC.
//    These tests will therefore be prone to false negatives.
//
// The first issue is a problem regardless of the test environment. If that
// were the only problem it would be fine to verify the behavior of these
// tests. (We do this with TEST_P, which has the same issue.)
//
// The second issue is specific to GCC prior to GCC 14. Being an environmental
// issue (e.g., results with GCC and Clang will differ), we need to handle it
// in a way that allows this test to pass for everyone.
//
// Therefore, we disable RGT output testing for TYPED_TEST and TYPED_TEST_P,
// unless GTEST_DEBUG_RGT is enabled. This allows users to optionally exercise
// these tests and make sure no Bad Stuff happens, but by default users aren't
// affected.

#if GTEST_DEBUG_RGT

using TypedTestTypes = testing::Types<char, int>;

template <typename T>
class NotRottenTypedTest : public testing::Test {};
TYPED_TEST_SUITE(NotRottenTypedTest, TypedTestTypes);

TYPED_TEST(NotRottenTypedTest, Passes) { EXPECT_EQ(0, TypeParam()); }

// If the assertions are conditional on the parameter, they aren't all
// executed in each instantiation. Because we treat these like helpers,
// and all the instantiations are deduplicated for reporting purposes,
// we report at the end of the test, and it's clean. If we support
// typed tests properly, we should continue not reporting until all
// instantiations have executed, to avoid false positives.

TYPED_TEST(NotRottenTypedTest, ConditionalOnType) {
  if (std::is_same<TypeParam, char>::value) {
    EXPECT_EQ(0, 0);
  } else {
    EXPECT_EQ(1, 1);
  }
}

template <typename T>
class IsRottenTypedTest : public testing::Test {};
TYPED_TEST_SUITE(IsRottenTypedTest, TypedTestTypes);

TYPED_TEST(IsRottenTypedTest, SkippedIsSadlyRotten) {
  assert(gtest_test_info_ == nullptr);
  if (testing::internal::AlwaysTrue()) {
    GTEST_SKIP();
  }
  EXPECT_EQ(0, 1);
}

TYPED_TEST(IsRottenTypedTest, DISABLED_DisabledIsSadlyRotten) {
  EXPECT_EQ(0, 1);
}

TYPED_TEST(IsRottenTypedTest, ActuallyRotten) {
  if (testing::internal::AlwaysFalse()) {
    EXPECT_EQ(0, 0);
  }
}

// RGT and Typed Parameterized Tests.

// See the note above about general issues with TYPED_TEST_P.

template <typename T>
class NotRottenTypeParamTest : public testing::Test {};
TYPED_TEST_SUITE_P(NotRottenTypeParamTest);

TYPED_TEST_P(NotRottenTypeParamTest, Passes) { EXPECT_EQ(0, TypeParam()); }

// If the assertions are conditional on the parameter, they aren't all
// executed in each instantiation. Because we treat these like helpers,
// and all the instantiations are deduplicated for reporting purposes,
// we report at the end of the test, and it's clean. If we support
// typed tests properly, we should continue not reporting until all
// instantiations have executed, to avoid false positives.

TYPED_TEST_P(NotRottenTypeParamTest, ConditionalOnType) {
  if (std::is_same<TypeParam, char>::value) {
    EXPECT_EQ(0, 0);
  } else {
    EXPECT_EQ(1, 1);
  }
}

REGISTER_TYPED_TEST_SUITE_P(NotRottenTypeParamTest,
                            Passes, ConditionalOnType);
INSTANTIATE_TYPED_TEST_SUITE_P(My, NotRottenTypeParamTest,
                               TypedTestTypes);

template <typename T>
class IsRottenTypeParamTest : public testing::Test {};
TYPED_TEST_SUITE_P(IsRottenTypeParamTest);

TYPED_TEST_P(IsRottenTypeParamTest, SkippedIsSadlyRotten) {
  if (testing::internal::AlwaysTrue()) {
    GTEST_SKIP();
  }
  EXPECT_EQ(0, 1);
}

TYPED_TEST_P(IsRottenTypeParamTest, DISABLED_DisabledIsSadlyRotten) {
  EXPECT_EQ(0, 1);
}

TYPED_TEST_P(IsRottenTypeParamTest, ActuallyRotten) {
  if (testing::internal::AlwaysFalse()) {
    EXPECT_EQ(0, 0);
  }
}

REGISTER_TYPED_TEST_SUITE_P(IsRottenTypeParamTest,
                            SkippedIsSadlyRotten,
                            DISABLED_DisabledIsSadlyRotten,
                            ActuallyRotten);
INSTANTIATE_TYPED_TEST_SUITE_P(My, IsRottenTypeParamTest,
                               TypedTestTypes);

#endif // GTEST_DEBUG_RGT

#endif // GTEST_HAS_RGT

File: googletest/test/gtest_rgt_output_test.py
#!/usr/bin/env python
#
# (c)2023 Sony Interactive Entertainment Inc
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#     * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above
# copyright notice, this list of conditions and the following disclaimer
# in the documentation and/or other materials provided with the
# distribution.
#     * Neither the name of Sony Interactive Entertainment Inc. nor the
# names of its contributors may be used to endorse or promote products
# derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

r"""Tests output of RGT detection for Google C++ Testing and Mocking Framework.

To update the golden file:
gtest_rgt_output_test.py --build_dir=BUILD/DIR --gengolden
where BUILD/DIR contains the built gtest_rgt_output_test_ file.
"""

import difflib
import os
import re
import sys
from googletest.test import gtest_test_utils


# The flag for generating the golden file
GENGOLDEN_FLAG = '--gengolden'
CATCH_EXCEPTIONS_ENV_VAR_NAME = 'GTEST_CATCH_EXCEPTIONS'

IS_LINUX = os.name == 'posix' and os.uname()[0] == 'Linux'
IS_WINDOWS = os.name == 'nt'

GOLDEN_NAME = 'gtest_rgt_output_test_golden_lin.txt'

PROGRAM_PATH = gtest_test_utils.GetTestExecutablePath('gtest_rgt_output_test_')

COMMAND_LIST_TESTS = ({}, [PROGRAM_PATH, '--gtest_list_tests'])
COMMAND_WITH_COLOR = ({}, [PROGRAM_PATH, '--gtest_color=yes'])
COMMAND_WITH_RGT_PASSING = ({}, [PROGRAM_PATH,
                                 '--gtest_treat_rotten_as_pass'])

GOLDEN_PATH = os.path.join(gtest_test_utils.GetSourceDir(), GOLDEN_NAME)


def ToUnixLineEnding(s):
  """Changes all Windows/Mac line endings in s to UNIX line endings."""

  return s.replace('\r\n', '\n').replace('\r', '\n')


def RemoveFileLocations(test_output):
  """Removes all file location info from a Google Test program's output.

  Args:
       test_output:  the output of a Google Test program.

  Returns:
       output with all file location info (in the form of
       'DIRECTORY/FILE_NAME:LINE_NUMBER: 'or
       'DIRECTORY\\FILE_NAME(LINE_NUMBER): ') replaced by
       'FILE_NAME:LINE_NUMBER: '.
  """

  test_output = re.sub(r'.*[/\\](gtest_main.cc)', r'\1', test_output)
  return re.sub(r'.*[/\\]((gtest_rgt_output_test_|gtest).cc)[:(](\d+)\)?\: ',
                r'\1:\3: ', test_output)


def ObscureLineNumbers(output):
  """Removes line numbers for error, Skipped and Failure notices.

  Args:
       output:  the output of a Google Test program.

  Returns:
       output with all file location info (in the form of
       'FILE_NAME:LINE_NUMBER: (error|Failure|Skipped)' replaced by
       'FILE_NAME:#: (error|Failure|Skipped)'.
  """

  return re.sub(r'((gtest_rgt_output_test_|gtest).cc):\d+: (error|Failure|Skipped)',
                r'\1:#: \3', output)


def RemoveTime(output):
  """Removes all time information from a Google Test program's output."""

  return re.sub(r'\(\d+ ms', '(? ms', output)


def NormalizeToCurrentPlatform(test_output):
  """Normalizes platform specific output details for easier comparison."""

  if IS_WINDOWS:
    # Removes the color information that is not present on Windows.
    test_output = re.sub('\x1b\\[(0;3\d)?m', '', test_output)
    # Changes failure message headers into the Windows format.
    test_output = re.sub(r': Failure\n', r': error: ', test_output)
    # Changes file(line_number) to file:line_number.
    test_output = re.sub(r'((\w|\.)+)\((\d+)\):', r'\1:\3:', test_output)

  return test_output


def RemoveTestCounts(output):
  """Removes test counts from a Google Test program's output."""

  output = re.sub(r'\d+ tests?, listed below',
                  '? tests, listed below', output)
  output = re.sub(r'\d+ FAILED TESTS',
                  '? FAILED TESTS', output)
  output = re.sub(r'\d+ tests? from \d+ test cases?',
                  '? tests from ? test cases', output)
  output = re.sub(r'\d+ tests? from ([a-zA-Z_])',
                  r'? tests from \1', output)
  return re.sub(r'\d+ tests?\.', '? tests.', output)


def NormalizeOutput(output):
  """Normalizes output (the output of gtest_rgt_output_test_.exe)."""

  output = ToUnixLineEnding(output)
  output = RemoveFileLocations(output)
  output = ObscureLineNumbers(output)
  output = RemoveTime(output)
  return output


def GetShellCommandOutput(env_cmd):
  """Runs a command in a sub-process, and returns its output in a string.

  Args:
    env_cmd: The shell command. A 2-tuple where element 0 is a dict of extra
             environment variables to set, and element 1 is a string with
             the command and any flags.

  Returns:
    A string with the command's combined standard and diagnostic output.
  """

  # Spawns cmd in a sub-process, and gets its standard I/O file objects.
  # Set and save the environment properly.
  environ = os.environ.copy()
  environ.update(env_cmd[0])
  p = gtest_test_utils.Subprocess(env_cmd[1], env=environ)

  return p.output


def GetCommandOutput(env_cmd):
  """Runs a command and returns its output with all file location
  info stripped off.

  Args:
    env_cmd:  The shell command. A 2-tuple where element 0 is a dict of extra
              environment variables to set, and element 1 is a string with
              the command and any flags.
  """

  # Disables exception pop-ups on Windows.
  environ, cmdline = env_cmd
  environ = dict(environ)  # Ensures we are modifying a copy.
  environ[CATCH_EXCEPTIONS_ENV_VAR_NAME] = '1'
  return NormalizeOutput(GetShellCommandOutput((environ, cmdline)))


def GetOutputOfAllCommands():
  """Returns concatenated output from several representative commands."""

  return (GetCommandOutput(COMMAND_WITH_COLOR) +
          GetCommandOutput(COMMAND_WITH_RGT_PASSING))


test_list = GetShellCommandOutput(COMMAND_LIST_TESTS)
SUPPORTS_RGT = 'Rotten' in test_list

CAN_GENERATE_GOLDEN_FILE = (SUPPORTS_RGT)

class GTestOutputTest(gtest_test_utils.TestCase):
  def testOutput(self):
    output = GetOutputOfAllCommands()

    golden_file = open(GOLDEN_PATH, 'rb')
    # A mis-configured source control system can cause \r appear in EOL
    # sequences when we read the golden file irrespective of an operating
    # system used. Therefore, we need to strip those \r's from newlines
    # unconditionally.
    golden = ToUnixLineEnding(golden_file.read().decode())
    golden_file.close()

    # We want the test to pass regardless of certain features being
    # supported or not.

    # We still have to remove type name specifics in all cases.
    normalized_actual = NormalizeToCurrentPlatform(output)
    normalized_golden = NormalizeToCurrentPlatform(golden)

    if CAN_GENERATE_GOLDEN_FILE:
      self.assertEqual(normalized_golden, normalized_actual,
                       '\n'.join(difflib.unified_diff(
                           normalized_golden.split('\n'),
                           normalized_actual.split('\n'),
                           'golden', 'actual')))
    else:
      normalized_actual = NormalizeToCurrentPlatform(
          RemoveTestCounts(normalized_actual))
      normalized_golden = NormalizeToCurrentPlatform(
          RemoveTestCounts(normalized_golden))

      # This code is very handy when debugging golden file differences:
      if os.getenv('DEBUG_GTEST_OUTPUT_TEST'):
        with open(os.path.join(
            gtest_test_utils.GetSourceDir(),
            '_gtest_rgt_output_test_normalized_actual.txt'), 'wb') as f:
          f.write(normalized_actual.encode())
        with open(os.path.join(
            gtest_test_utils.GetSourceDir(),
            '_gtest_rgt_output_test_normalized_golden.txt'), 'wb') as f:
          f.write(normalized_golden.encode())

      self.assertEqual(normalized_golden, normalized_actual)


if __name__ == '__main__':
  if GENGOLDEN_FLAG in sys.argv:
    if CAN_GENERATE_GOLDEN_FILE:
      output = GetOutputOfAllCommands()
      golden_file = open(GOLDEN_PATH, 'wb')
      golden_file.write(output.encode())
      golden_file.close()
    else:
      message = (
          """Unable to write a golden file when compiled in an environment
that does not support all the required features (rotten green tests).
Please build this test and generate the golden file using Blaze on Linux.""")

      sys.stderr.write(message)
      sys.exit(1)
  else:
    gtest_test_utils.Main()
